verilog/tools/ls/autoexpand.cc

```diff
@@ -243,27 +362,265 @@ void AutoExpander::Module::EmitPortConnections(
+    return port.declaration == Port::Undeclared &&
+           port.direction == Port::Input;
+  });
+}
+
+void AutoExpander::Module::EmitUndeclaredInoutDeclarations(
+    std::ostream &output, const absl::string_view indent) const {
+  EmitPortBodyDeclarations(output, indent, [](const Port &port) {
+    return port.declaration == Port::Undeclared &&
+           port.direction == Port::Inout;
+  });
+}
+
+void AutoExpander::Module::EmitUndeclaredOutputDeclarations(
+    std::ostream &output, const absl::string_view indent) const {
+  EmitPortBodyDeclarations(output, indent, [](const Port &port) {
+    return port.declaration == Port::Undeclared &&
+           port.direction == Port::Output;
+  });
+}
+
+void AutoExpander::Module::EmitUndeclaredOutputWireDeclarations(
+    std::ostream &output, const absl::string_view indent,
+    const absl::string_view auto_span) const {
+  absl::flat_hash_set<absl::string_view> declared_wires;
+  for (const auto &reg : FindAllNetVariables(symbol)) {
+    const SyntaxTreeLeaf *const net_name_leaf =
+        GetNameLeafOfNetVariable(*reg.match);
+    const absl::string_view net_name = net_name_leaf->get().text();
+    if (net_name.end() <= auto_span.begin() &&
+        net_name.begin() >= auto_span.end()) {
+      declared_wires.insert(net_name);
+    }
+  }
+  const auto is_undeclared_connected_and_not_wire = [&](const Port &port) {
+    return port.declaration == Port::Undeclared && port.connection &&
+           !declared_wires.contains(port.name);
+  };
+
   for (const Port &port : ports_) {
-    if (!pred(port)) continue;
-    if (first) {
-      if (output.tellp() != 0) output << ',';
-      output << '\n' << indent << "// " << header;
-      first = false;
-    } else {
-      output << ',';
+    if (port.direction == Port::Inout &&
+        is_undeclared_connected_and_not_wire(port)) {
+      output << indent << "wire " << port.name << ";  // To/From "
+             << port.connection->instance << " of " << port.connection->type
+             << "\n";
+    }
+  }
+  for (const Port &port : ports_) {
+    if (port.direction == Port::Output &&
+        is_undeclared_connected_and_not_wire(port)) {
+      output << indent << "wire " << port.name << ";  // From "
+             << port.connection->instance << " of " << port.connection->type
+             << "\n";
     }
-    auto connected_name = port.name;
+  }
+}
+
+void AutoExpander::Module::EmitUnconnectedOutputRegDeclarations(
+    std::ostream &output, const absl::string_view indent,
+    const absl::string_view auto_span) const {
+  absl::flat_hash_set<absl::string_view> declared_regs;
+  for (const auto &reg : FindAllRegisterVariables(symbol)) {
+    const SyntaxTreeLeaf *const reg_name_leaf =
+        GetNameLeafOfRegisterVariable(*reg.match);
+    const absl::string_view reg_name = reg_name_leaf->get().text();
+    if (reg_name.end() <= auto_span.begin() &&
+        reg_name.begin() >= auto_span.end()) {
+      declared_regs.insert(reg_name);
```

Comment:
Does any unit test run into this condition ?

Reply:
I thought I had tested this, but it seems not. And the condition is wrong, of course.